Ensure the original grad_mode is restored during backward

albanD opened the pull request:
Fix #31715

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31884 Ensure the original grad_mode is restored during backward**

Differential Revision: [D19301076](https://our.internmc.facebook.com/intern/diff/D19301076)

albanD committed:
Ensure the original grad_mode is restored during backward

[ghstack-poisoned]

gchanan commented on test/test_autograd.py:
```diff
@@ -3351,6 +3351,30 @@ def test_checkpointing(self):
+        inp = torch.rand(3, requires_grad=True)
+
+        MyFunction.apply(inp).sum().backward()
+        MyFunction.apply(inp).sum().backward(create_graph=True)
```
I'm guessing this is true, but just to check: this covers both original is True and False?

albanD replied:
Exactly.

albanD committed:
Update on "Ensure the original grad_mode is restored during backward"

Fix #31715


Differential Revision: [D19301076](https://our.internmc.facebook.com/intern/diff/D19301076)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 35df89be:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=35df89bed6a1b3f1f8db0e92b03fe6cc1d2b2e67).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4158399) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=81297455) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=81297455))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
 
  git checkout -b <new-branch-name> 
 
HEAD is now at 35df89bed6 Update on "Ensure the original grad_mode is restored during backward" 
+ git reset --hard 35df89bed6a1b3f1f8db0e92b03fe6cc1d2b2e67 
HEAD is now at 35df89bed6 Update on "Ensure the original grad_mode is restored during backward" 
+ git merge --allow-unrelated-histories --no-edit --no-ff ca72df06ae2f90ec0243094d8127f1b190fcea5f 
Auto-merging torch/csrc/autograd/engine.cpp 
Auto-merging test/test_autograd.py 
CONFLICT (content): Merge conflict in test/test_autograd.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD committed:
Update on "Ensure the original grad_mode is restored during backward"

Fix #31715


Differential Revision: [D19301076](https://our.internmc.facebook.com/intern/diff/D19301076)

[ghstack-poisoned]

albanD committed:
Update on "Ensure the original grad_mode is restored during backward"

Fix #31715


Differential Revision: [D19301076](https://our.internmc.facebook.com/intern/diff/D19301076)

[ghstack-poisoned]
